versioned_docs/version-4.4/building-extensions/plugins/basic-content-plugin.md

```diff
@@ -0,0 +1,291 @@
+use Joomla\CMS\Plugin\PluginHelper;
+use Joomla\DI\Container;
+use Joomla\DI\ServiceProviderInterface;
+use Joomla\Event\DispatcherInterface;
+use My\Plugin\Content\Shortcodes\Extension\Shortcode;
+
+    return new class() implements ServiceProviderInterface
+    {
+        public function register(Container $container)
+        {
+            $container->set(
+                PluginInterface::class,
+                function (Container $container) {
+    
+                    $config = (array) PluginHelper::getPlugin('content', 'shortcodes');
+                    $subject = $container->get(DispatcherInterface::class);
+                    $app = Factory::getApplication();
+                    
+                    $plugin = new Shortcode($subject, $config);
+                    $plugin->setApplication($app);
+    
+                    return $plugin;
+                }
+            );
+        }
+    };
+```
+This is pretty much boilerplate code for obtaining your plugin from the Dependency Injection Container, and you need to change just 3 lines from what seems to be standard code in Joomla core plugins.
+```php
+use My\Plugin\Content\Shortcodes\Extension\Shortcode;
+```
+Ensure that this aligns with your `<namespace>` in the manifest file and your `namespace` statement and class name in the Extension class file. .
+
+```php
+$config = (array) PluginHelper::getPlugin('content', 'shortcodes');
+```
+Ensure that this contains your plugin type and element, matching the manifest file.
+
+```php
+$plugin = new Shortcode($subject, $config);
+```
+Ensure that this matches your class in your `src/Extension` directory.
+
+## Extension Class
+This is the main code of the plugin. Hopefully the comments in the code explain what is going on.
+
+As explained in [Joomla 4 and 5 changes](./joomla-4-and-5-changes), code which triggers the Events can use a `GenericEvent` or a concrete Event, eg `ContentPrepareEvent`. In both these cases you can get the arguments using
+```php
+[$context, $article, $params, $page] = array_values($event->getArguments());
+```
+but you have to check in your code how to return the result. 
+
+Using this approach means that you don't need to change your plugin code if the code which is triggering the event changes from using a generic Event class to a specific event class. 
+
+**plg_shortcodes/src/Extension/Shortcode.php**
+```php
+<?php
+namespace My\Plugin\Content\Shortcodes\Extension;
+
+// no direct access
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Plugin\CMSPlugin;
+use Joomla\Event\Event;
+use Joomla\Event\SubscriberInterface;
+use Joomla\CMS\Factory;
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Event\Result\ResultAwareInterface;
+
+class Shortcode extends CMSPlugin implements SubscriberInterface
+{
+    public static function getSubscribedEvents(): array
+    {
+        // we want this plugin to work only on the site, not eg on administrator or api
+        if (Factory::getApplication()->isClient('site')) {

```

Comment:
Please do not use it here, or better remove this example, it is a bad one :smile: 
I know Joomla have it in the core, but it not a good use.

In general the method `getSubscribedEvents` should not depend from current runtime.

Reply:
Alternatively add it in to the listeners like:
```php
if ($this->getApplication()->isClient('site')) {
  return;
}
```

Reply:
@Fedik @HLeithner Thanks for taking the time to review! I'm just rather surprised at the suggestion to move the client check from getSubscribedEvent. Surely it's more performant to do it there, so that Joomla doesn't waste time triggering events that are going to be ignored anyway, isn't it? 

If there's a good reason to do it in the listener then it would be worth adding an explanation in the document as to why it should be done there instead.

Reply:
What I mean, we should not to advertise a bad practice in our documentation, even if Joomla use it in own code :wink: 

It do not take much time to trigger the event, I would not wory about it. 
This not realy about perfomance but more about code reliability.

When in future we find a better soluton, then we can update the manual. But for now better not to advertise such thing.

Reply:
The root in how the events works, the listener initialisation:
```php
$subscriber = new EventSubscriber(); // Our plugin
$dispatcher->addSubscriber($subscriber); // Call $subscriber->getSubscribedEvents() internally
```
This code will initialise new Subscriber quickly.

Keeping the plugin `__constructor` and `getSubscribedEvents` light will keep whole booting process fast, and code more reliable.

In fact, not all listeners from subscriber is called in the end. 
Everything depend from plugin group and current runtime, example `onContentPrepare` will never be called if current component do not need it. Or `onBeforeCompileHead` will never be called if request not `html` and so on.

Reply:
ok, I've changed that. Agreed it would be slightly faster for the site (which is the most important), even though it would be a bit slower for the administrator in having to add the listeners.

Reply:
Thanks!